### PR TITLE
feat(core): add $dynamicRef / $dynamicAnchor support for OpenAPI 3.1

### LINKED ISSUE #3352: feat: support $dynamicRef / $dynamicAnchor (JSON Schema 2020-12 / OpenAPI 3.1)
## Summary

Orval parses OpenAPI 3.1 specs containing JSON Schema 2020-12 `$dynamicRef` / `$dynamicAnchor`, but generated TypeScript currently loses dynamic reference semantics and falls back to `unknown` in cases that can be resolved from the active dynamic scope.

This affects recursive schemas and reusable wrapper patterns where preserving concrete SDK types matters.

## Reproduction

Validator-backed fixtures:

- Recursive category tree: https://github.com/aqeelat/openapi-dynamicref-adoption-tracker/blob/main/fixtures/recursive-category-tree.yaml
- Nested workspace resources: https://github.com/aqeelat/openapi-dynamicref-adoption-tracker/blob/main/fixtures/nested-workspace-resources.yaml

Generic-wrapper fixtures:

- Named schemas: https://github.com/aqeelat/openapi-dynamicref-adoption-tracker/blob/main/fixtures/generic-schema-binding.yaml
- Inline response binding: https://github.com/aqeelat/openapi-dynamicref-adoption-tracker/blob/main/fixtures/paginated-response.yaml

## Expected Behavior

Generated SDK types should preserve concrete dynamic-reference types when the active dynamic scope provides an explicit binding.

Unresolved or unsupported `$dynamicRef` cases should remain safe and fall back to `unknown`.

## Implementation Tasks

- [ ] Core implementation (#3353)
  - Support explicit dynamic-scope bindings:
    - schema-owned `$dynamicAnchor`
    - `$defs` entries with `$dynamicAnchor` + schema `$ref`
    - unbound `$defs` anchors as generic type parameters
  - Emit reusable TypeScript generics and bound aliases.
  - Support inline response bindings and `allOf` intersections.
  - Fall back to `unknown` for unresolved or unsupported anchors.
  - Add documentation.
  - Add samples.
  - Add unit and integration tests.

- [ ] Enhancement: lexical fallback for unbound `$dynamicRef`
  - When a `$dynamicRef` has no override in the current dynamic scope, fall back to the schema that originally declared the `$dynamicAnchor` (lexical position), per JSON Schema 2020-12.
  - Likely requires changing `dynamicScope` from a flat per-schema map to a stack-based structure threaded through the full resolution chain (`resolveValue` → `getScalar` → object resolution).
  - Example: `Lizard` has `$dynamicRef: '#Pet'` but no `$dynamicAnchor: Pet` override — currently falls back to `unknown`, should resolve to `Pet` per spec.

- [ ] Enhancement: external `$dynamicRef` support
  - Investigate external `$dynamicRef` support separately.
  - Keep current behavior as safe fallback to `unknown` until external resolution is implemented.

- [ ] Enhancement: non-schema component dynamic bindings, if needed
  - PR #3353 only binds `$defs` refs pointing to `#/components/schemas/...`.
  - Decide whether `responses`, `parameters`, or `requestBodies` should ever participate in dynamic-scope resolution.

## Stability

The feature can be considered stable after the base implementation PR lands because the core behavior is explicit-binding-only and does not rely on temporary config flags.

Follow-up tasks can expand lexical fallback, external-reference support, and any non-schema component binding support without changing the core stability contract.

## Update (2026-05-18)

PR #3353 was adjusted based on review feedback:

- Removed the proposed temporary opt-in flag.
- Removed global/static anchor fallback.
- Kept explicit dynamic-scope support for schema-owned anchors, `$defs` schema bindings, generic template parameters, bound aliases, inline response bindings, and `allOf` intersections.

This keeps the implementation closer to JSON Schema dynamic-scope semantics and avoids shipping an incomplete config surface.

### PR DESCRIPTION
**Pull request by aqeelat**
## Summary

Adds support for JSON Schema 2020-12 `$dynamicRef` / `$dynamicAnchor` in OpenAPI 3.1 schemas, resolving #3352.

The implementation is always active when these keywords are present. There is no config flag.

## What's included

- `isDynamicReference()` type guard + `OpenApiDynamicReferenceObject`
- `getDynamicAnchorName()` parser for fragment-only dynamic refs
- `buildDynamicScope()` for explicit dynamic-scope bindings from:
  - a schema's own `$dynamicAnchor`
  - `$defs` entries with `$dynamicAnchor` + `$ref` to `#/components/schemas/...`
  - unbound `$defs` entries used as generic type parameters
- `resolveDynamicRef()` for resolving anchors through the current `context.dynamicScope`
- `extractBoundAliasInfo()` for reusable generic template bindings
- Generic interface emission for reusable templates
- Type alias emission for bound generic usages
- Inline response binding support
- `allOf` bound alias support via intersection types
- `dynamicAnchorsToUniqueParamNames()` for deterministic deduplication when sanitized anchor names collide (e.g. `foo-bar` and `foo_bar` both sanitize to `foo_bar`)
- Partially-bound alias support — unbound template slots are passed through as generic parameters with correct arity
- Documentation for supported patterns and limitations
- Samples for recursive schemas and paginated/generic response wrappers
- Unit and integration coverage for recursive schemas, generic wrappers, inline response bindings, `allOf` intersections, non-identifier anchors, external refs, null `$defs` entries, scoped cache isolation, anchor collisions, and partially-bound aliases

## Supported patterns

| Pattern | Example | Result |
|---|---|---|
| Self-referential `$dynamicAnchor` | `BaseCategory.children: $dynamicRef('#category')` | recursive concrete type |
| Generic template via unbound `$defs` anchor | `PaginatedTemplate` with `$defs.itemType.$dynamicAnchor` | `interface PaginatedTemplate<itemType>` |
| Bound generic alias | `$defs.itemType.$ref` points to `#/components/schemas/Pet` | `type PaginatedPets = PaginatedTemplate<Pet>` |
| Multi-parameter templates | `Template<folderType, resourceType>` | ordered generic type arguments |
| Partially-bound aliases | bind `itemType` to `User`, leave `cursorType` unbound | `type Alias<cursorType> = Template<User, cursorType>` |
| Colliding anchor names | `foo-bar` and `foo_bar` in same `$defs` | `foo_bar`, `foo_bar2` — deterministic suffixes |
| Inline response bindings | response schema has `$ref` + `$defs` binding | concrete response type |
| `allOf` bound aliases | bound template plus extra schema (including `$ref` siblings) | `Template<Args> & Extra` |
| External `$dynamicRef` | `other.json#anchor` | falls back to `unknown` |

## Explicitly not included

This PR does not infer bindings by scanning sibling schemas in `components.schemas`.

A `$dynamicRef` only resolves when the active dynamic scope contains an explicit matching anchor binding. Unresolved anchors fall back to `unknown`.

Non-schema component refs in dynamic-scope bindings (for example `#/components/responses/...`) are not bound because dynamic resolution emits schema refs only.

## Limitations

- Each schema in `components.schemas` is generated once with a single dynamic scope. If the same named component is referenced from multiple call sites with different `$defs` bindings, fully call-site-aware emission is not attempted here.
- External `$dynamicRef` documents are not fetched and fall back to `unknown`.
- Inline `$defs` entries without `$ref` are treated as generic type parameters, not concrete anonymous bindings.
- Lexical fallback for unbound `$dynamicRef` anchors remains follow-up work tracked in #3352.

Closes #3352

## Summary by CodeRabbit

* **New Features**
  * Automatic OpenAPI 3.1 $dynamicRef / $dynamicAnchor support with polymorphic generics and bound-alias instantiation
  * New generated SDK samples demonstrating dynamic-ref and paginated-response patterns (Axios-based clients, typed models, paginated envelopes)

* **Documentation**
  * Added reference docs with examples, expected fallbacks, and known limitations

* **Tests**
  * Extensive new tests and snapshots covering dynamic scope, resolution, edge cases, collisions, and regressions

* **Bug Fixes**
  * Improved scope isolation, cache safety, and null-safe handling to avoid cross-scope leaks

[Review Change Stack](https://app.coderabbit.ai/change-stack/orval-labs/orval/pull/3353?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by melloware** [comment]
@soartec-lab @anymaniax would like input here on the naming of the config etc.  Does this need to be an override switch?

**Comment by aqeelat** [comment]
> @soartec-lab @anymaniax would like input here on the naming of the config etc. Does this need to be an override switch?

@melloware 
There are two ways to go about it:
1. do the entire feature in one PR which means no need for the flag. The PR will have many clean easy to review commits, but it will be hard to review in one go.
2. do it in multiple PRs, keep the flag, then remove the flag once all tasks are done.

I have a few optimization tweaks that I need to do. but then we can remove the flag.

**Comment by aqeelat** [comment]
Converted to draft because apparently there’s more work that I need to do. Will update later. But the question still stands on whether you guys are comfortable with a big PR or if you rather have smaller PRs. But this decision can be done later once we have a full context.

**Comment by soartec-lab** [comment]
Hi @aqeelat I think we need to make small changes in stages. This will allow the scope to gradually expand. Since we're making improvements in parts, there's no need for a "development in progress" flag that will be removed at the end. Also, it would be very helpful if you could write out the plan as an issue so we can track the progress.

**Comment by aqeelat** [comment]
@coderabbitai review

**Comment by aqeelat** [comment]
@coderabbitai full review

File: packages/core/src/resolvers/ref.ts [hunk 1]
```diff
@@ -189,6 +190,112 @@ function getSchema<TSchema extends object = OpenApiComponentsObject>(
...
+    const refInfo = getRefInfo(
+      `#/components/schemas/${name.replaceAll('~', '~0').replaceAll('/', '~1')}`,
+      context,
+    );
+
+    return {
+      name: refInfo.name,
+      schemaName: refInfo.originalName,
+    };
+  };
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/resolvers/ref.ts:389]: fixed.

[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/resolvers/ref.ts:389]: Skipping this one.

`encodeJsonPointerSegment()` is only used for component schema names, and OpenAPI component keys are constrained by `^[a-zA-Z0-9._-]+$`. Characters like `%`, `/`, `~`, and `#` are not valid component schema keys, so the URI-fragment encoding concern does not represent a valid OpenAPI scenario for these call sites.

For valid component keys, the current helper is sufficient. Broader raw `$ref` encoding behavior would be a separate concern outside this `$dynamicRef` change.

File: packages/core/src/resolvers/value.ts [hunk 2]
```diff
@@ -19,13 +19,131 @@ interface ResolveValueOptions {
...
+  'contains',
+  'else',
+  'if',
+  'items',
+  'not',
+  'propertyNames',
+  'then',
+  'unevaluatedItems',
+  'unevaluatedProperties',
+] as const;
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/resolvers/value.ts:35]: Skipping this one.

For OpenAPI 3.1 / JSON Schema 2020-12, tuple-position schemas are represented by `prefixItems`, and `prefixItems` is already included in `schemaArrayKeys`.

`items` is a single schema in this dialect, so keeping it in `schemaObjectKeys` is intentional. Changing it to array handling would imply support for an older/non-target tuple shape and would make this traversal less aligned with OAS 3.1 semantics.

File: packages/core/src/writers/schemas.ts [hunk 3]
```diff
@@ -218,18 +218,13 @@ function normalizeCanonicalImportPaths(
...
-      // `relative` derives from canonical.importPath (built via getPath) so it
-      // normally ends with `fileExtension`; the `.ts$` branch is a defensive
-      // fallback for legacy/hardcoded `.ts` paths that don't match the
-      // configured fileExtension (e.g. `.gen.ts`).
-      const withoutFileExtension = relative.endsWith(fileExtension)
-        ? relative.slice(0, -fileExtension.length)
-        : relative.replace(/\.ts$/, '');
-      const importPath = `${withoutFileExtension}${importExtension}`;
+      const importPath = relativePath.endsWith(fileExtension)
+        ? `${relativePath.slice(0, -fileExtension.length)}${importExtension}`
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/writers/schemas.ts]: Deferring this one to a follow-up.

This may be valid, but it affects generated import-path/module-resolution behavior more broadly than `$dynamicRef`. I'd rather cover it with a focused regression test for the relevant `fileExtension` / `tsconfig` combination instead of bundling it into this dynamic-ref review pass.

File: samples/dynamic-ref/api/petstore.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/**
+ * Generated by orval v8.10.0 🍺
+ * Do not edit manually.
+ * Petstore with dynamic references
+ * Demonstrates OpenAPI 3.1 `$dynamicRef` / `$dynamicAnchor` for polymorphic schema dispatch.
+The `Pet` base schema defines a `$dynamicAnchor` that is overridden by `Cat` and `Dog`.  Each schema has a `playmates` property using `$dynamicRef: '#Pet'` — which resolves to `Cat[]` inside Cat, and `Dog[]` inside Dog.
+`Lizard` does NOT override the anchor, so its `playmates` falls back to `unknown` since only explicit bindings are resolved.
+
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply samples/dynamic-ref/api/petstore.ts]: Deferring this one.

The generated header formatting issue is pre-existing and not specific to `$dynamicRef`. Fixing it would likely touch many generated snapshots/samples, so it should be handled separately to keep this PR focused on dynamic-ref behavior.

**Comment by aqeelat** [comment]
@coderabbitai resume

File: packages/core/src/getters/ref.ts [hunk 5]
```diff
@@ -80,3 +85,12 @@ export function getRefInfo($ref: string, context: ContextSpec): RefInfo {
     refPaths,
   };
 }
+
+export function getDynamicAnchorName(dynamicRef: string): string {
+  if (!dynamicRef.startsWith('#')) {
+    throw new Error(
+      `$dynamicRef with external document is not supported: ${dynamicRef}`,
+    );
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/getters/ref.ts]: Agreed this is a minor cleanup opportunity, but removing the check would change semantics for edge cases where the caller doesn't pre-check. Deferring to keep the diff focused.

File: packages/core/src/resolvers/ref.ts [hunk 6]
```diff
@@ -189,6 +190,112 @@ function getSchema<TSchema extends object = OpenApiComponentsObject>(
...
+        const anchorName = (defSchema as Record<string, unknown>)
+          .$dynamicAnchor as string;
+        const refInDef = (defSchema as OpenApiReferenceObject).$ref;
+        if (refInDef) {
+          const { name, originalName } = getRefInfo(refInDef, context);
+          scope[anchorName] = { name, schemaName: originalName };
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/resolvers/ref.ts]: By design. Inline schemas in `$defs` without a `$ref` are treated as generic type parameters (unbound anchors), not concrete bindings. Only `$defs` entries with both `$dynamicAnchor` and `$ref` create bindings.

File: samples/dynamic-ref/api/petstore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (aqeelat) [reply samples/dynamic-ref/api/petstore.ts]: fixed in #3393

[the remark this answers is not in the document]
Comment (aqeelat) [reply samples/dynamic-ref/api/petstore.ts]: fixed in #3393

**Comment by aqeelat** [comment]
@coderabbitai resume
@coderabbitai full review

File: samples/paginated-response/vitest.snapshots.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { defineConfig } from 'vitest/config';
+import pkg from './package.json' with { type: 'json' };
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply samples/paginated-response/vitest.snapshots.ts:2]: ### Not valid — no action taken

The finding claims `test.name` is an invalid Vitest option and should be removed or replaced with `testNamePattern`. However, this project uses Vitest **4.0.18**, and `test.name` is a documented, valid configuration option that accepts either a string or an object of shape `{ label: string, color?: string }` for project naming.

The pattern used in this file:

```ts
name: { label: pkg.name },
```

is directly shown in the official Vitest 4.x docs at [`guide/projects.md`](https://github.com/vitest-dev/vitest/blob/v4.0.7/docs/guide/projects.md):

```ts
{
  test: {
    name: { label: 'node', color: 'green' },
    environment: 'node',
  }
}
```

`testNamePattern` serves a completely different purpose (filtering tests by name at runtime) and is not a replacement for `test.name` (which identifies a project in a workspace). The property is valid and should remain as-is.

File: samples/paginated-response/vitest.snapshots.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+
+export default defineConfig({
+  test: {
+    name: { label: pkg.name },
+    include: ['api-generation.spec.ts'],
+    silent: 'passed-only',
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply samples/paginated-response/vitest.snapshots.ts:8]: ### Not valid — no action taken

The finding's guidance to "keep `test.silent` as-is" is correct, but for completeness: while the TypeScript type definition in Vitest declares `silent: boolean`, the runtime CLI option parser explicitly accepts the string value `'passed-only'`:

```js
// vitest/dist/chunks/cac.DVeoLl0M.js
if (value === "passed-only") return value;
```

This is a known type looseness in Vitest itself — the CLI help text documents `'passed-only'` as a valid value: *"Silent console output from tests. Use `'passed-only'` to see logs from failing tests only."* The configuration works correctly at runtime and requires no changes.

File: packages/core/src/generators/schema-definition.ts [hunk 9]
```diff
@@ -159,6 +166,31 @@ function shouldCreateInterface(schema: OpenApiSchemaObject) {
...
+    ) {
+      const anchor = rec.$dynamicAnchor as string;
+      params.push({
+        anchorName: anchor,
+        paramName: dynamicAnchorToParamName(anchor),
+      });
+    }
+  }
+  return params;
+}
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/generators/schema-definition.ts:192]: The collision (`item-id` and `item_id` both sanitizing to `item_id`) is theoretically possible but not a practical concern:
1. **Already consistent.** Both `collectGenericParams` and `buildDynamicScope` call `dynamicAnchorToParamName` independently on the same anchor, so they produce identical names for the same anchor — no drift.
2. **Unrealistic spec.** A collision requires a single schema's `$defs` to contain two *different* `$dynamicAnchor` values that differ only in dash vs underscore. `$dynamicAnchor` names are unique binding identifiers within a schema — no real-world spec would have this.
3. **Disproportionate complexity.** Fixing this would require computing a shared anchor→paramName mapping once and threading it through both functions (called independently at 4 sites in `generateSchemaDefinitions`), all to defend against an authoring error that won't occur in practice.

**Comment by aqeelat** [comment]
@coderabbitai resolve

**Comment by aqeelat** [comment]
@melloware @snebjorn @soartec-lab @anymaniax 

The PR is now ready. I know the PR is huge but I was unable to break it down any further. I suggest using the CodeRabbit change stack to review the changes.

Sorry for the notification spam 😅

**Comment by aqeelat** [comment]
@coderabbitai resume

**Comment by melloware** [comment]
@aqeelat can you rebase and fix the merge conflicts?

File: packages/core/src/generators/schema-definition.ts [hunk 10]
```diff
@@ -184,18 +216,113 @@ function generateSchemaDefinitions(
     ];
   }
 
+  const alias = extractBoundAliasInfo(schema, context);
+  if (alias) {
+    const genericParams = alias.genericParams.map((paramName) => ({
+      anchorName: paramName,
+      paramName,
+    }));
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/generators/schema-definition.ts:224]: Valid observation but no live bug:

1. The original anchor name is already lost upstream — `extractBoundAliasInfo` at `ref.ts:282` pushes only the deduplicated `paramName` into `genericParams: string[]`.
2. The `{ anchorName, paramName }` objects at `schema-definition.ts:221-224` are only consumed at line 227 via `p.paramName` (for the generic suffix). `anchorName` is **set but never read**.
3. The potential `dynamicScope` divergence concern (deduplicating only alias-scope unbound anchors vs. all template anchors) could produce mismatched param names for colliding anchors — but this only matters if `extraSchemas` contain `$dynamicRef` references to those anchors, which no current code path exercises.

File: packages/core/src/generators/dynamic-ref.test.ts [hunk 11]
```diff
@@ -0,0 +1,950 @@
...
+                $ref: '#/components/schemas/User',
+              },
+              foo_bar: {
+                $dynamicAnchor: 'foo_bar',
+                $ref: '#/components/schemas/Group',
+              },
+            },
+            $ref: '#/components/schemas/DualSlotTemplate',
+          },
+        },
```
[the remark this answers is not in the document]
Comment (aqeelat) [reply packages/core/src/generators/dynamic-ref.test.ts:875]: The `$ref` paths (`#/components/schemas/User`, `#/components/schemas/Group`) are resolved purely by string extraction via `getRefInfo()` (`ref.ts:380-389`) — the actual schema objects are never looked up. `extractBoundAliasInfo` and `buildDynamicScope` both derive type names from the ref path, not from the schema definition. Adding `User`/`Group` definitions would be cosmetic.

**Review by melloware — APPROVED** [review]
